### PR TITLE
BUG: Restore setting TargetPoints after ReachMode

### PR DESCRIPTION
**Pull request by blowekamp**
Check that the target reach mode and expected number of target point are value in VerifyPreconditions instead of when setting expected number of targets.

This restores prior behavior of being able to set the target reach mode before setting the target points.

## PR Checklist
- [ ] No [API changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#breaking-changes) were made (or the changes have been approved)
- [ ] No [major design changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#design-changes) were made (or the changes have been approved)
- [ ] Added test (or behavior not changed)
- [ ] Updated API documentation (or API not changed)
- [ ] Added [license](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Utilities/KWStyle/ITKHeader.h) to new files (if any)
- [ ] Added Python wrapping to new files (if any) as described in [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) Section 9.5
- [ ] Added [ITK examples](https://github.com/InsightSoftwareConsortium/ITKSphinxExamples) for all new major features (if any)

Refer to the [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) for
further development details if necessary.

### CODE REVIEW

**Comment by blowekamp** [comment]
This regression was reported in SimpleITK/SimpleITK#1801.

The PR is based on the release branch, and it is intended that this change will be merged into the release branch as well.

**Review by dzenanz — APPROVED** [review]
Looks good. In commit message some rephrasing might be needed. E.g. "number of target point are value in" -> "number of target point**s** are **correct** in".

File: Modules/Filtering/FastMarching/include/itkFastMarchingUpwindGradientImageFilter.hxx [hunk 1]
```diff
@@ -59,6 +59,26 @@ FastMarchingUpwindGradientImageFilter<TLevelSet, TSpeedImage>::PrintSelf(std::os
   os << indent << "Target value: " << m_TargetValue << std::endl;
 }
 
+template <typename TLevelSet, typename TSpeedImage>
+void
+FastMarchingUpwindGradientImageFilter<TLevelSet, TSpeedImage>::VerifyPreconditions() ITKv5_CONST
+{
+  Superclass::VerifyPreconditions();
+
+  switch (this->m_TargetReachedMode)
```
Comment (jhlegarreta) [inline Modules/Filtering/FastMarching/include/itkFastMarchingUpwindGradientImageFilter.hxx:68]: Is the `this` pointer necessary here? We usually do no require it.

File: Modules/Filtering/FastMarching/include/itkFastMarchingUpwindGradientImageFilter.h [hunk 2]
```diff
@@ -240,7 +237,7 @@ class ITK_TEMPLATE_EXPORT FastMarchingUpwindGradientImageFilter : public FastMar
    *  Returns true if at least a target point exists; returns false otherwise.
    */
   bool
-  IsTargetPointsExistenceConditionSatisfied()
+  IsTargetPointsExistenceConditionSatisfied() const
```
Comment (jhlegarreta) [inline Modules/Filtering/FastMarching/include/itkFastMarchingUpwindGradientImageFilter.h:240]: To me, this is not related to the bug itself, and I'd advocate for adding it in a separate commit.

File: Modules/Filtering/FastMarching/include/itkFastMarchingUpwindGradientImageFilter.h [hunk 3]
```diff
@@ -258,7 +255,7 @@ class ITK_TEMPLATE_EXPORT FastMarchingUpwindGradientImageFilter : public FastMar
    *  Raises an exception if the conditions are not satisfied.
    */
   void
-  VerifyTargetReachedModeConditions(unsigned int targetModeMinPoints = 1)
+  VerifyTargetReachedModeConditions(unsigned int targetModeMinPoints = 1) const
```
Comment (jhlegarreta) [inline Modules/Filtering/FastMarching/include/itkFastMarchingUpwindGradientImageFilter.h:258]: Same as above.

**Review by jhlegarreta — COMMENTED** [review]
Thank for doing this Brad. I'm sorry this has caused a failure in SimpleITK

> This restores prior behavior of being able to set the target reach mode before setting the target points.

Not sure if I get this: you still need the target points (or the number of target points) to verify the pre-conditions.

Unrelated to the patch set, I realize that the this `enum` has survived the effort Hans pioneered to convert them to class enums:
https://github.com/InsightSoftwareConsortium/ITK/blob/9306f48136bacc7b8bf17f5cd4f0f693658299dc/Modules/Filtering/FastMarching/include/itkFastMarchingUpwindGradientImageFilter.h#L195

Edit:
> Unrelated to the patch set, I realize that the this `enum` has survived the effort Hans pioneered to convert them to class enums:

 PR #3863.

File: Modules/Filtering/FastMarching/include/itkFastMarchingUpwindGradientImageFilter.h (hunk 2 above)
Comment (blowekamp) [reply Modules/Filtering/FastMarching/include/itkFastMarchingUpwindGradientImageFilter.h:240]: VerifyPreconditions is a const member function and it calls this method, so it needed to be changed.

**Comment by blowekamp** [comment]
@jhlegarreta I am going ahead and merging this PR, so that the enum PR can be rebased onto master and conflicts avoided.
